time-based-packet-transmission-offload.md

```diff
@@ -153,6 +171,29 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
 
 ```
 
+To obtain the `NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO` structure, a driver should call the `NET_BUFFER_LIST_INFO` macro with an _Id of `TcpIpTxTimeNetBufferListInfo`
+```C
+typedef struct _NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO {
+  struct {
+    ULONG IsIPv4 : 1;
+    ULONG IsIPv6 : 1;
+    ULONG IsTcp : 1;
+    ULONG IsUdp : 1;
+    ULONG TxTime : 64;
```

Comment:
ULONG 64bit?

Reply:
Also, I don't think we can use more than 32 bits in this struct. Each NBL OOB info field is actually a PVOID, which can only fit 32 bits.

Reply:
We should scope this offload to 64-bit systems only if that reduces our OOB complexity. I can't imagine these offloads being relevant for 32 platforms.

Reply:
The only 32-bit platforms MSFT ships now are x86 for Client SKU. Server is all 64-bit. I'd be perfectly happy scoping to 64-bit if we need to.

Reply:
I am not sure what this TxTime represents. If it takes 64 bits, we still can't fit it in a NBL oob.

Reply:
It is time an associated packet to be sent.

Reply:
I know it's time but what is the reference and unit?

Reply:
I'm investigating how windows get time, but Linux set txtime from userspace as bellow
```C
	struct timespec ts;
	int cnt, err;
	__u64 txtime;

	clock_gettime(clkid, &ts);

	/* Start one to two seconds in the future. */
	ts.tv_sec += 1;
	ts.tv_nsec = 1000000000 - waketx_delay;
	normalize(&ts);

	txtime = ts.tv_sec * 1000000000ULL + ts.tv_nsec;
	txtime += waketx_delay;
```